### PR TITLE
Implement Sky Shaders

### PR DESCRIPTION
**Pull request by clayjohn**
First of all, huge thanks to @BastiaanOlij. He started the work and got the foundation completely ready. 

This implements the sky shader proposal (Pre GIP) documented here: https://docs.google.com/document/d/1rOvUvKrLgNN4kdT52nKvvWpW6i102BEPKM_t0sk5AfY/edit?usp=sharing

**What is it?**
Sky shaders add a new type of shader ``shader_type sky``. They are attached to ``Sky`` resources to define how the sky background and radiance cubemaps are drawn. Sky Shaders remove the need for ``PanoramaSky`` and ``ProceduralSky`` and replace them with materials that can be attached to the ``Sky`` resource. To start I have ``PanoramaSkyMaterial`` (just takes a single panorama texture and applies it), ``ProceduralSkyMaterial`` (same settings as ProceduralSky), and ``PhysicalSkyMaterial`` (defines an earth-like sky with a single sun based on the Preetham atmospheric scattering model).

**Why?**
Sky Shaders allow much more flexible sky behaviour. The largest hurdle they overcome is the bulky radiance map generation from Godot 3.x. In Godot 3.x, there is no way to have a custom sky with animated properties with proper reflections. With Sky Shaders, reflections are updated automatically when needed. 

Sky shaders also make it possible for users to write their own shaders. For instance, if a user wants to add real time clouds to their project, or if they want to create a special type of sky.

**The Juicy Details!**
Sky shaders have two main built-in variables ``EYEDIR`` and ``COLOR``. ``COLOR`` is the final output color of the shader, while ``EYEDIR`` is the input direction in world space. Using these two variables a wide range of effects are possible. 

Sky Shaders automatically check if any directional lights are in the scene. The first 4 directional lights can be accessed from the shader with the ``LIGHTX_**`` properties. ``LIGHT0_ENABLED``, ``LIGHT0_DIRECTION``, ``LIGHT0_ENERGY`` and ``LIGHT0_COLOR``.

Shaders can be animated with ``TIME``

The camera position in world space is ``POSITION``

Built in are two optional subpasses which can be accessed with ``HALF_RES_TEXTURE`` and ``QUARTER_RES_TEXTURE``. These subpasses run the sky shader on a half-resolution or quarter-resolution texture to allow expensive calculations to be done fewer times (e.g. for clouds). Currently, to use the subpasses you must set the appropriate render mode ``use_half_res_pass`` or ``use_quarter_res_pass``. This behaviour could be made automatic whenever ``**_TEXTURE`` is used. but reduz thought it would be better to have render modes.

The UVs of the screen rect are in ``SCREEN_UV`` which can be used to access the textures.

To specify different behaviour for each pass, you can use the ``AT_**_PASS`` built-ins. for example
```
if (AT_CUBEMAP_PASS) {
  COLOR = vec3(1.0); // reflections will be all white
} else if (AT_HALF_RES_PASS) {
  vec4 col = generate_fancy_clouds(EYEDIR, TIME); // clouds will be rendered to half res texture
  COLOR = col.rgb;
  ALPHA = col.a; // subpasses can use alpha
} else {
  COLOR = texture(HALF_RES_TEXTURE, SCREEN_UV).rgb; // background will read from HALF_RES_PASS
}
```

**Sky Materials**

There are 3 pre-made materials for users to choose from. Ideally these cover most use-cases, but where they are not enough, users can create their own shaders.

All ``**SkyMaterials`` can be converted to ``ShaderMaterial``s if users want to copy them and tweak them.

_PanoramaSkyMaterial_: behaves exactly like ``PanoramaSky``

_ProceduralSkyMaterial_: behaves almost the same as ``ProceduralSky`` except sun properties are drawn from the first 4 directional lights and it updates instantly and so is viable for realtime skies.

_PhysicalSkyMaterial_: sky material controlled by various physical properties. It can only have 1 sun, so it uses the first ``DirectionalLight`` in the scene. It is based on the Preetham model, therefore it sacrifices physical accuracy for speed and flexibility. I added the ability to change the scattering wavelengths, ground color, and sun disk size. The Preetham model is a daylight model, so it just goes black at nightime, we can consider adding a physically based night model as well in the future.

**Examples**
_PhysicalSkyMaterial at sunset_
Screenshot (98)

_ShaderMaterial: shader ported from shadertoy_
Screenshot (104)

_ProceduralSkyMaterial: beware the programmers lack of color-sense_
Screenshot (105)

### CODE REVIEW

File: servers/visual/shader_types.cpp [hunk 1]
```diff
@@ -285,7 +285,43 @@ ShaderTypes::ShaderTypes() {
...
 	shader_modes[VS::SHADER_PARTICLES].modes.push_back("keep_data");
 
+	/************ SKY **************************/
+
+	shader_modes[VS::SHADER_SKY].functions["fragment"].built_ins["COLOR"] = ShaderLanguage::TYPE_VEC3;
+	shader_modes[VS::SHADER_SKY].functions["fragment"].built_ins["ALPHA"] = ShaderLanguage::TYPE_FLOAT;
+	shader_modes[VS::SHADER_SKY].functions["fragment"].built_ins["EYEDIR"] = constt(ShaderLanguage::TYPE_VEC3);
+	shader_modes[VS::SHADER_SKY].functions["fragment"].built_ins["SCREEN_UV"] = constt(ShaderLanguage::TYPE_VEC2);
+	shader_modes[VS::SHADER_SKY].functions["fragment"].built_ins["SKY_COORDS"] = constt(ShaderLanguage::TYPE_VEC2);
+	shader_modes[VS::SHADER_SKY].functions["fragment"].built_ins["TIME"] = constt(ShaderLanguage::TYPE_FLOAT);
```
Comment (Chaosus) [inline servers/visual/shader_types.cpp]: This should be replaced by `shader_modes[VS::SHADER_SKY].functions["global"].built_ins["TIME"] = constt(ShaderLanguage::TYPE_FLOAT);`

**Comment by akien-mga** [comment]
From the CI logs there seems to be at least one `-Wsign-compare` warning: https://travis-ci.org/github/godotengine/godot/jobs/664693277?utm_medium=notification&utm_source=github_status, and some missing stubs in the dummy rasterizer: https://travis-ci.org/github/godotengine/godot/jobs/664693281?utm_medium=notification&utm_source=github_status

**Comment by tinmanjuggernaut** [comment]
Super awesome!

> All **SkyMaterials can be converted to ShaderMaterials if users want to copy them and tweak them.

Excellent!

> ProceduralSkyMaterial: behaves almost the same as ProceduralSky except sun properties are drawn from the first 4 directional lights and it updates instantly and so is viable for realtime skies.

So it draws suns in the sky based upon your directional lights? And matches color, angle, intensity, etc?

Is size of the drawn sun independent from energy? e.g. can I do a large dim orange sun, or a small intense blue sun with the DL and Sky settings? 

Does light come from the sky still? Previously the light/sun acted like an under powered HDRI. If it uses the DL does it also get light from the sky and drawn sun?

Can I not draw a sun? e.g. use a directional light for sky illumination, without it appearing in the sky.

Was sky contribution fixed? Like reducing the calculation to 1/3 or 1/2 so it matches HDRs closer?

On the 3.x ProceduralSky there was a dot that appeared at the zenith of the sky sphere. Did you happen to discover and fix that? I never mentioned it anywhere.

**Comment by clayjohn** [comment]
> So it draws suns in the sky based upon your directional lights? And matches color, angle, intensity, etc?

Yep!

> Is size of the drawn sun independent from energy? e.g. can I do a large dim orange sun, or a small intense blue sun with the DL and Sky settings?

Yep!

> Does light come from the sky still? Previously the light/sun acted like an under powered HDRI. If it uses the DL does it also get light from the sky and drawn sun?

Yep, skies have always behaved the same as HDRIs. It is up to you how weak or strong the light is. 

> Can I not draw a sun? e.g. use a directional light for sky illumination, without it appearing in the sky.

Not with the built in material. You will have to edit it yourself to get that.

> Was sky contribution fixed? Like reducing the calculation to 1/3 or 1/2 so it matches HDRs closer?

I'm not sure what you mean. So probably not. 

> On the 3.x ProceduralSky there was a dot that appeared at the zenith of the sky sphere. Did you happen to discover and fix that? I never mentioned it anywhere.

I did not notice it until you mentioned it here. But the dot indeed has gone away

**Comment by tinmanjuggernaut** [comment]
All sounds good. Looking forward to trying it out. Screenshots look amazing!

> > Was sky contribution fixed? Like reducing the calculation to 1/3 or 1/2 so it matches HDRs closer?
> 
> I'm not sure what you mean. So probably not.

I'm refering to [this proposal discussion](https://github.com/godotengine/godot-proposals/issues/348), specifically under the first post, "My Proposal / Ambient Light"

> At 1 [sky contribution], procedural sky is unnaturally blue. It appears far stronger than it is with HDRIs - maybe reduce the ProcSky scale to 1/3rd. Some HDRIs look fine at 1. So I'd either default the settings for both at 3/0.3, or reduce the internal ProcSky Sky Contrib scale down to 33-50% of what it is now, and boost energy x3, then you could leave both HDRI Sky Contrib and ProcSky Sky Contrib at 1 and 1.

You [commented](https://github.com/godotengine/godot-proposals/issues/348#issuecomment-571644579):

> Regarding the blue tint of the white objects see this tweet: https://twitter.com/matkovskid/status/1214561009794347009?s=20

> In nature, shaded objects appear very blue, due to the sky accounting for most of their light. Whereas it is only within direct sunlight that they appear their proper color. Similarly here, when a direct light is applied, whites appear properly white. 

And I responded right after that post, but essentially, that twitter pic is an alpine shot. The left image is an extreme case with intense blue in the shadow. The right image doesn't have so much blue in the shadow. 

SkyContribution is a fine feature, it's just way too strong. The blue from ProceduralSKy is about 2x-3x stronger than it is from HDRIs. If the calculation were brought down, it would be on par with PanoramaSky.

**Comment by clayjohn** [comment]
@tinmanjuggernaut I think you are confusing a few things. sky_contribution is a setting that allows you to blend a custom ambient light color with the sky light. In real life, there is no sky_contribution slider. The image I sent was of the same mountain, one angle was in complete shadow, the other was in direct light. Similarly in Godot, when in complete shadow, pure white objects appear much too blue. The solution is to add a directional light, not to hack in a custom ambient color. At any rate, its not connected to this PR whatsoever, we can discuss over on your proposal if you would like to discuss further :). 

PS: I tried to tweak the ProceduralSkyMaterial to use colors closer to what you proposed. Please give this PR a try and see what you think. I would love to start it off with the best default values possible.

@akien-mga and all interested. Just force pushed a change to finish off the sky shaders. Should be ready for testing now. Hopefully it passes CI now. I also took the chance to clean up the radiance map generation codebase now that we arent using panorama textures directly anymore.

File: servers/visual/shader_types.cpp (hunk 1 above)
Comment (clayjohn) [reply servers/visual/shader_types.cpp]: Thanks. Added :)

**Comment by clayjohn** [comment]
@akien-mga looks like CI is having a difficult time.

**Comment by clayjohn** [comment]
First pass of the docs are up.

Looks like CI is having trouble starting the Mac build. But it doesn't seem related to this PR. 

This PR is ready for testing by anyone interested :)

**Comment by BastiaanOlij** [comment]
Awesome work @clayjohn , I'll find some time to play with this during the week.

**Comment by Dimev** [comment]
Looks really good!
It would be nice to also have a way to do fog with shaders (I know it's currently possible) similar to how the sky shaders work.

The proposal states this is to be decided, but I can see it being useful as it would allow realistic atmospheric fog.

**Comment by clayjohn** [comment]
@Dimev fog is being ported over in https://github.com/godotengine/godot/pull/36989

If you want to do custom fog with a shader, you can already do it in a spatial shader. 

It's not possible to have a dedicated fog shader as fog needs to be calculated at the same time as the rest of the object. 

Hope that helps!

**Review by reduz — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks a ton!

**Comment by Lexpartizan** [comment]
I would like to talk about the problems that I encountered when developing my shader 
https://github.com/Lexpartizan/Godot_sky_shader of the sky
and what I miss. Firstly, I would like to be able to convert Physical Sky Material to usual sky shader for edit code and add features.

If we talk about Physical Sky Material, then just the color of the sky and the Sun is not enough for any game. The sky needs clouds and the moon. That is, either I have to personally change the Physical Sky Material code, or it would be better to add the ability to simply specify textures for clouds, star_field and the Moon.
This should not complicate the code much (if there is a texture, just mix it, you can also add mixing to this texture depending on the strength of light or the height of the Sun), but it will make Physical Sky Material much more usable.
In addition, I would like to render the clouds separately in the hotel shader and save them to a texture or viewport. Because the texture of the clouds can come in handy for other effects, for example, God_Rays.

Now about Panorama Sky. 
I would like to have examples of converting coordinates for a projection to a panorama and back. For example, now I can't correctly display the textured moon on the panorama in any position. Because I'm not good at math :-)
I would like to be able to use the hemisphere, not the sphere. To increase pixel density at the same resolution. This is important for clouds. Volumetric clouds noticeably slow down at a resolution of 4k, and at a resolution of 1k, pixelation is already noticeable closer to the horizon.

**Comment by Zireael07** [comment]
@Lexpartizan: Many simpler games will get just fine without clouds and moon (and the moon can be pretty easily faked by reusing the sun, just paler and weaker)

However, regarding converting coordinates to panorama and back you have a very good point.

**Comment by Lexpartizan** [comment]
The fact is that if you specify a simple texture as the Shader, it does not complicate the Shader at all. Just the sky will be mixed with alpha = 0.0 and there will be no changes. But those who need clouds can simply specify either a static texture or a viewport of the Shader with clouds.

**Comment by tinmanjuggernaut** [comment]
Adding a texture lookup to the shader will slow it down for all who don't use it. For those who need it, the existing shader can be converted to a ShaderMaterial that you can edit. Then you can easily add a texture lookup and a corresponding uniform so you can add it in the editor. Maybe two lines of code. One for the uniform, and one for the lookup and mix.

As for the flat to equirectangular conversion, the source for your repo (danil) has that conversion in "Dynamic sky and reflection" already. You'll have to analyze it a bit to pull it out, but the code is already written in your repo to go one way. It's not much work to reverse the equation to go the opposite way.

**Comment by Calinou** [comment]
@tinmanjuggernaut Can the texture lookup be made optional? I'd expect Godot to be able to skip it if the shader does not define any overlay texture.

**Comment by Lexpartizan** [comment]
> Adding a texture lookup to the shader will slow it down for all who don't use it.

Atmospheric scaterring not very fast shader. And this add only one mix instruction, becouse if texture empty, we have vec4(0). Attenuation clouds depending from Sun height and enough two instruction for clamp and multiply.

**Comment by Zireael07** [comment]
He means texture lookup, i.e. TEXTURE() call.

**Comment by Lexpartizan** [comment]
> TEXTURE() call.

Understand. But I think it's worth it. Again, atmosphering scaterring is not the fastest algorithm.

**Comment by tinmanjuggernaut** [comment]
@lexpartizan Someone who won't use it, or is making a low end / mobile game, will be stuck rendering 1-3 extra lookups per pixel. At 1080p, when looking at the sky, that's 2M-6M slow operations that they can't get rid of! That could easily cut the framerate in half or more, even on a high end card. Texture lookups can be costly. 2-3 just for clouds seems like too many. 

@calinou In the shader you'd have to have a variable that told it to render the texture and a trinary ()? to conditionally add the lookup. As I understand it, most compilers will make two versions and be fine, but I've also read about some low end compilers/hardware tanking on branching.

You probably mean in the C++ code, where you could rewrite the shader you're sending for compilation based upon if the textures are used. Since you're parsing anyway, this is probably the better way to go. 

Your suggestion is certainly a good one that would resolve both sides of the topic. I would also certainly like to have clouds on systems that can handle it.

**Comment by Lexpartizan** [comment]
Thank you for explaining. Yes, the solution with bool flags seems to me a good compromise.

**Comment by Calinou** [comment]
> As I understand it, most compilers will make two versions and be fine, but I've also read about some low end compilers/hardware tanking on branching.

This is Vulkan, I'd be surprised if any hardware that supports it would run into trouble with static branching.

**Comment by clayjohn** [comment]
This discussion should be moved over to a Godot Proposal. Currently, the provided sky materials just handle sun and daylight sky. Anything else should be in a proper proposal format that can be discussed.

**Comment by Lexpartizan** [comment]
https://github.com/godotengine/godot-proposals/issues/835

**Comment by Zireael07** [comment]
I know this is a bit of a necro, but could the code for the demo shaders shown in the blog/OP (the clouds at half pass and the ported shadertoy shader) be posted anywhere? I managed to create full pass clouds, but I'm struggling to use the half pass... either the docs are out of date or I'm doing something wrong.

**Pull request merged**